src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php

```diff
@@ -160,6 +160,22 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             // transformation is merged back into the original collection
             $builder->addEventSubscriber(new MergeCollectionListener(true, true));
         }
+
+        // To avoid some problem against treating of array (e.g. Array to string conversion),
+        // we have to first ensure the all elements of submitted data ain't an array.
+        $builder->addEventListener(FormEvents::PRE_SUBMIT, function (FormEvent $event) {
+            $data = $event->getData();
+
+            if (!is_array($data)) {
+                return;
+            }
+
+            foreach ($data as $v) {
+                if (null !== $v && !is_string($v)) {
```

Comment:
The listener you mentioned is only added when choice is multiple. Even though we would do so we still need to a new listener for this.